### ISSUE TITLE
[FEATURE]: A better, more informative, append-only logger

### ISSUE THREAD

**Issue by eysi09** [garden-io/garden#3254]
# Feature Request
## Background / Motivation
The Garden Core logger currently has a few issues:

* It’s a little buggy:
  * The spinner goes out of whack if there's a lot of output.
* It could be a lot more informative:
  * The default log level shows too little information in terms of what’s happening in the user’s system. You might know that Garden is building a given module but can’t see what’s happening in the build. If it e.g. takes longer than expected, you’ll need to re-run the command with a higher log level to see what actually happened (assuming you’re not using Cloud).
  * When using the basic logger or on higher log levels you don’t see the context of the log lines. You might see something like: `#29 [built 1/1] RUN yarn run build` but can’t tell what build it belongs to.
  * Sections are confusing. Sometimes they refer to internal things and sometimes to stack graph nodes with out making it clear what the type of the node is.
  * When you install a Helm chart, Helm will print documentation about the chart from the maintainer. E.g. deprecation warnings. Garden doesn't show this information.
* Error messages are confusing:
  * We sometimes get complaints about error messages being hard to parse, in particular from developers.

## Suggested Implementation(s)
Below are some _suggestions_ for improvements. We should make sure to experiment with this and see what makes sense and what doesn’t. In particular for larger projects and focusing on the non-happy path.

* **Make the logger append-only**
  * We should completely remove the `setState` functionality and stop editing lines in place. Rather, each new log line should be, well, a new line.
  * We can have a single status line at the bottom with a spinner when in watch mode that makes it clear when Garden is responding to events.
* **Informative sections**
  * Sections should map directly to stack graph nodes, showing both the type and name of the node. E.g. `build.api`. We can reserve a keyword for internal sections, e.g.: `[internal] graph`
  * Sections should display the name of any external tool Garden is shelling out to along with the command name when applicable. E.g. `build.api ⇒ docker build ⇒ …` or  `deploy.api => helm install => ...`
  * All log lines should have sections. So instead of seeing `#29 [built 1/1] RUN yarn run build` you see `build.api ⇒ docker build ⇒ #29 [built 1/1] RUN yarn run build`.
* **Hints**
  * We could maintain a map of hints that we apply if applicable. Attaching info about external tools or providers to the error context could be helpful here. In pseudo code, something like:

```tsx
if (error.detail?.tool === "helm" && error.message.match(/timedout waiting for contition/) {
  log(hints.HELM_TIME_OUT_ERROR)
}
```

* **Err on being too informative**
  * Previously we’ve hidden most of what’s happening “under the hood” from the user so as not to flood the terminal. I think we should reverse this philosophy and basically default to showing what’s happening. This includes streaming build logs, and test and task results as they are written to stdout.
  * We should add a command line option to suppress these types of outputs (except on failure).
  * We should add a config option to configure log level for each individual node. You might e.g. choose to always suppress the build logs for your worker service but always show them for your API service.
    * This could be configure at the config level and overwritten at the command line level.
    * Having a user specific runtime config would be helpful here.
    * Future work: Custom failure / success messages. Could e.g. be used to print ingress links when Garden can’t detect them.
  * We should always print the names of commands that Garden is running when it starts them.
  * Always show durations when applicable.
  * Show which build mode is being used.
* **Misc**
  * Add empty lines between logs from different sections for easier readability.
  * Default to showing timestamps?
  * Do not indent sections?
  * In general, be thoughtful and intentional about error messages :slightly_smiling_face:

### Example
Again, this is just a _suggestion_, based on the points above. We should definitely play around and see what makes sense IRL.

```bash
➜  garden deploy
Deploy 🚀 

━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
🌍  Running in namespace eysi in environment dev 
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

[11:34:43] ℹ [internal] garden-cloud   →  Authorizing...
[11:34:43] ✔ [internal] garden-cloud   →  Successfully authorized!
[11:34:43] ℹ [internal] garden-cloud   →  Visit project at https://staging.enterprise.garden.io/projects/3
[11:34:43] ℹ [internal] garden-cloud   →  Initializing...
[11:34:43] ✔ [internal] garden-cloud   →  Ready!

[11:34:43] ℹ [internal] providers      →  Getting status...
[11:34:43] ✔ [internal] providers      →  Finished getting status (took 6 sec)

[11:34:43] ℹ [internal] graph          →  Resolving 33 modules...
[11:34:43] ✔ [internal] graph          →  Finished resolving modules (took 5.5 sec)

[11:34:43] ℹ build.api-image           →  Getting build status for v-84c1a8af97... 

[11:34:43] ℹ build.frontend            →  Getting build status for v-9999999999... 

[11:34:43] ℹ build.api-image           →  No image with ID v-84c1a8af97 found. Will build using build mode 'cluster-buildkit'.

[11:34:43] ✔ build.frontend            →  Image already built!

[11:34:43] ℹ build.api-image           →  Running command in build kit pod: "docker build [...]"  
[11:34:43] ℹ build.api-image           →  [docker build] →  #1 [internal] load build definition from api.dockerfile

[11:34:43] ℹ deploy.frontend           →  Running command "kubectl apply [...]"

[11:34:43] ℹ build.api-image           →  [docker build] →  #1 transferring dockerfile: 2.82kB done
[11:34:43] ℹ build.api-image           →  [docker build] →  #1 DONE 0.0s
[11:34:43] ℹ build.api-image           →  [docker build] →  #2 [internal] load .dockerignore
[11:34:43] ℹ build.api-image           →  [docker build] →  #2 transferring context: 2B done 

[11:34:43] ℹ deploy.frontend           →  [kubectl apply] →  Waiting for resources to be ready...
[11:34:43] ℹ deploy.frontend           →  [kubectl apply] →  Status of Service api is "ready"

[11:34:43] ℹ build.api-image           →  [docker build] →  #2 DONE 0.0s 
[11:34:43] ℹ build.api-image           →  [docker build] →  #4 [built 1/1] RUN yarn run build
[11:34:43] ℹ build.api-image           →  [docker build] →  #5 exporting cache
[11:34:43] ℹ build.api-image           →  [docker build] →  #5 preparing build cache for export 0.1s done
[11:34:43] ℹ build.api-image           →  [docker build] →  #5 DONE 0.1s
[11:34:43] ✔ build.api-image           →  Finsished building version v-84c1a8af97 of api-image (took 35 sec)

[11:34:43] ℹ deploy.frontend           →  [kubectl apply] →  Resources ready
[11:34:43] ✔ deploy.frontend           →  Finished deploying frontend (took 10 sec)

[11:34:43] ℹ deploy.api                →  Running command "kubectl apply [...]"
[11:34:43] ℹ deploy.api                →  [kubectl apply] →  Warning Failed kubelet 27s (x4 over 62s) gke-xxx-yyy Error: ImagePullBackOff
[11:34:43] ✖ deploy.api                →  Error (after 5 sec). See below:

  An error occurred when Garden ran the following command:

  /path/to/kubectl apply -f [...] 

  Here\'s the full error:

  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
  Warning Failed kubelet 27s (x4 over 62s) gke-xxx-yyy Error: ImagePullBackOff
  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ℹ Garden hint:

  This looks like an "ImagePullBackOff" error from "Kubernetes". This can happen when the cluster isn't
  authorized to pull the image from container registry.

  Check out this section of our docs for how to configure image pulls: https://docs.garden.io/pulling-images

  If you think this hint is wrong, please open an issue using this link:
  https://github.com/garden-io/garden/issues/new?labels=bad_hint&template=BAD_HINT.md&title=%5BBad hint for %5D%3A+ImagePullBackOff

Done! Failed with 1 error.
```

## How important is this feature for you/your team?
:rose: It’s a nice to have, but nice things are nice :slightly_smiling_face:

## Relevant issues
* #3538

**Comment by eysi09** [comment]
Writing this up here to get the conversation going and solicit feedback from others. 

We can already start experimenting with this from the v0.13 branch. One idea would be to just create a dummy command that spits out a bunch of logs that we can use for iterating on this. 

Once we v0.13 is merged we'll have done a lot of the ideation and validation and can quickly implement.

**Comment by worldofgeese** [comment]
> * When you install a Helm chart, Helm will print documentation about the chart from the maintainer. E.g. deprecation warnings. Garden doesn't show this information.

I believe this feature in particular would be a huge win for Garden as these chart success messages will often contain critical information like connection URIs, accessible ports, first-time setup instructions, etc.

**Comment by eysi09** [comment]
I'm going to go ahead and close this since the first iteration is merged on 0.13. 

We need to create a follow up issue for the "hints database".
